File: JAVA/s/src/main/java/s/Node.java
package s;

import s.exp.Exp;

public class Node<T> {
	private Node(T v,Node<T> n) {
		this.value=v;
		this.next=n;
		if(n!=null) {
			this.length=n.Length()+1;
		}else {
			this.length=1;
		}
	}
	private int length;
	public int Length() {
		return length;
	}
	private T value;
	private Node<T> next;
	public T First() {return value;}
	public Node<T> Rest() {return next;}

	//不换行，默认
	@Override
	public String toString() {
		StringBuilder sb=new StringBuilder();
		toString(sb);
		return sb.toString();
	}
	
	private static void toString(Object v,StringBuilder sb) {
		if(v==null) {
			sb.append("[]");
		}else if(v instanceof Node) {
			((Node)v).toString(sb);
		}else if(v instanceof String) {
			sb.append(mb.Util.string_to_trans(v.toString(),'"','"',null));
		}else if(v instanceof Function) {
			Function f=(Function) v;
			if(f instanceof Function.UserFunction) {
				((Function.UserFunction)f).toString(sb);
			}else
			if(f.ftype()==Function.Type.user) {
				/*优化但并非用户函数*/
				sb.append(f.toString());
			}else
			if(f.ftype()==Function.Type.buildIn) {
				/*内置函数*/
				sb.append("'").append(f.toString());
			}else
			if(f.ftype()==Function.Type.cache) {
				/*cache函数*/
				sb.append("[]");
			}
		}else if(v instanceof Boolean) {
			if((Boolean)v) {
				sb.append("true");
			}else {
				sb.append("false");
			}
		}else if(v instanceof Integer) {
			sb.append(v.toString());
		}else {
			String sx=v.toString();
			if(sx==null) {
				//某些内置库无法被序列化，比如match/cache-run
				sb.append("[]");
			}else {
				//js的toString只有字面值，在列表中需要转义，虽然用'比较有意义。
				//比如内置库
				sb.append("'").append(sx);
			}
		}
	}
	//嵌套不换行
	public void toString(StringBuilder sb) {
		sb.append("[ ");
		for(Node<T> t=this;t!=null;t=t.next) {
			toString(t.First(),sb);
			sb.append(" ");
		}
		sb.append("]");
	}
	
	public static <T> Node<T> extend(T x,Node<T> xs){
		return new Node<T>(x,xs);
	}
    /*反转*/
    public static <T> Node<T> reverse(Node<T> node) {
    	Node<T> r=null;
    	for(Node<T> t=node;t!=null;t=t.Rest()) {
    		r=extend(t.First(),r);
    	}
    	return r;
    }
	public static Node<Object> kvs_extend(String key,Object value,Node<Object> kvs){
		return extend(key,extend(value,kvs));
	}
	public static Object kvs_find1st(Node<Object> kvs,String key) {
    	if(kvs==null) {
    		return null;
    	}else {
    		Node<Object> r_kvs=kvs.Rest();
    		if(r_kvs==null) {
    			System.out.println("as:"+key+":"+kvs.toString());
    		}
    		if(key.equals(kvs.First())) {
    			return r_kvs.First();
    		}else {
    			return kvs_find1st(r_kvs.Rest(),key);
    		}
    	}
	}
}

File: C#/s/s/Node.cs
using System;
using System.Collections.Generic;
using System.Text;

namespace s
{
    public class Node<T>
    {
        private T first;
        private Node<T> rest;
        private int length;
        private Node(T first, Node<T> rest)
        {
            this.first = first;
            this.rest = rest;
            if (rest == null)
            {
                length = 1;
            }
            else
            {
                length = rest.Length() + 1;
            }
        }
        public T First() { return first; }
        public Node<T> Rest() { return rest; }
        public int Length() { return length; }


        public override string ToString()
        {
            StringBuilder sb = new StringBuilder();
            toString(sb);
            return sb.ToString();
        }
        protected void toString(StringBuilder sb)
        {
            sb.Append("[");
            toString(sb, first);
            for (Node<T> tmp = rest; tmp != null; tmp = tmp.Rest())
            {
                sb.Append(" ");
                toString(sb, tmp.First());
            }
            sb.Append("]");
        }
        /// <summary>
        /// 专供调用
        /// </summary>
        /// <param name="sb"></param>
        /// <param name="v"></param>
        private static void toString(StringBuilder sb, Object v)
        {
            if (v == null)
            {
                sb.Append("[]");
            }
            else if (v is Node<Object>)
            {
                (v as Node<Object>).toString(sb);
            }
            else if (v is String)
            {
                sb.Append(Util.stringToEscape(v as String, '"', '"', null));
            }
            else if (v is Function)
            {
                Function f = v as Function;
                if (f.Function_type() == Function.Function_Type.Fun_BuildIn)
                {
                    sb.Append("'");
                    sb.Append(v.ToString());
                }
                else if (f.Function_type() == Function.Function_Type.Fun_Better)
                {
                    sb.Append(v.ToString());
                }
                else if (f.Function_type() == Function.Function_Type.Fun_User)
                {
                    (f as UserFunction).toString(sb);
                }
                else if (f.Function_type() == Function.Function_Type.Fun_Cache)
                {
                    sb.Append("[]");
                }else
                {
                    sb.Append(v.ToString());
                }
            }
            else if (v is bool)
            {
                if ((bool)v)
                {
                    sb.Append("true");
                }
                else
                {
                    sb.Append("false");
                }
            }
            else if (v is int)
            {
                sb.Append(v.ToString());
            }
            else
            {
                String vx = v.ToString();
                if (vx == null)
                {
                    sb.Append("[]");
                }
                else
                {
                    sb.Append("'").Append(v.ToString());
                }
            }
        }

        public static Node<T> reverse(Node<T> list)
        {
            Node<T> r = null;
            while (list != null)
            {
                r = new Node<T>(list.First(), r);
                list = list.Rest();
            }
            return r;
        }

        /**************************一些库函数********************************/

        public static Node<T> extend(T value, Node<T> scope)
        {
            return new Node<T>(value, scope);
        }
        public static Node<T> list(params T[] ps)
        {
            return extends(ps, null);
        }
        public static Node<T> extends(T[] ps, Node<T> r)
        {
            for (int i = ps.Length-1; i >-1; i--)
            {
                r = new Node<T>(ps[i], r);
            }
            return r;
        }
        public static Node<Object> kvs_extend(String key, Object value, Node<Object> scope)
        {
            return new Node<Object>(key, new Node<Object>(value, scope));
        }
        public static Object kvs_find1st(Node<Object> kvs, String k)
        {
            if (kvs == null)
            {
                return null;
            }
            else
            {
                String key = (String)kvs.First();
                kvs = kvs.Rest();
                if (key == k)
                {
                    return kvs.First();
                }
                else
                {
                    return kvs_find1st(kvs.Rest(), k);
                }
            }
        }
    }
}
